### PR TITLE
CUDA: Main refactor

### PR DESCRIPTION
**Pull request by samueltootle**
This PR aims to address adding device (CUDA) support to `main_c` for the BHaH infrastructure.

### CODE REVIEW

File: nrpy/infrastructures/BHaH/MoLtimestepping/__init__.py [hunk 1]
```diff
@@ -0,0 +1 @@
+from . import MoL_allocators, MoL_gridfunction_names, MoL_register_all, MoL_rk_substep, MoL_step_forward, RK_Butcher_Table_Dictionary
```
Comment (zachetienne) [inline nrpy/infrastructures/BHaH/MoLtimestepping/__init__.py:1]: Let's format this file using `black`.

Comment (samueltootle) [reply nrpy/infrastructures/BHaH/MoLtimestepping/__init__.py:1]: Resolved in latest commits

File: nrpy/infrastructures/BHaH/tests/main_c___openmp.c [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+    // Step 5.d: Main loop, part 4: Step forward in time using Method of Lines with RK4 algorithm,
+    //           applying  boundary conditions.
+    MoL_step_forward_in_time(&commondata, griddata);
+
+    // Step 5.e: Main loop, part 5 (post_MoL_step_forward_in_time): Finish up step in time
+    // (nothing here; specify by setting post_MoL_step_forward_in_time string in register_CFunction_main_c().)
+
+  } // End main loop to progress forward in time.
+  BHAH_DEVICE_SYNC();
```
[the remark this answers is not in the document]
Comment (samueltootle) [reply nrpy/infrastructures/BHaH/tests/main_c___openmp.c:81]: Resolved in latest commits

File: nrpy/infrastructures/BHaH/main_c.py [hunk 3]
```diff
@@ -96,43 +131,88 @@ def register_CFunction_main_c(
...
+        if parallelization in ["cuda"]
+        else """Set up initial data.
 initial_data(&commondata, griddata);
 """
+    )
     allocate_storage_code = """Allocate storage for non-y_n gridfunctions, needed for the Runge-Kutta-like timestepping.
 for(int grid=0; grid<commondata.NUMGRIDS; grid++)
   MoL_malloc_non_y_n_gfs(&commondata, &griddata[grid].params, &griddata[grid].gridfuncs);
-"""
+""".replace(
```
Comment (zachetienne) [inline nrpy/infrastructures/BHaH/main_c.py:213]: Let's remove all the replace() stuff and instead set the variable 
`griddata_dev = "griddata_device" if parallelization in ["cuda"] else "griddata"`
at the top and use f-strings.
